source/server/configuration_impl.cc

```diff
@@ -83,11 +98,27 @@ void MainImpl::initialize(const envoy::config::bootstrap::v3::Bootstrap& bootstr
     server.listenerManager().addOrUpdateListener(listeners[i], "", false);
   }
 
-  stats_flush_interval_ =
-      std::chrono::milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(bootstrap, stats_flush_interval, 5000));
-
   initializeWatchdogs(bootstrap, server);
-  initializeStatsSinks(bootstrap, server);
+  initializeStatsConfig(bootstrap, server);
+}
+
+void MainImpl::initializeStatsConfig(const envoy::config::bootstrap::v3::Bootstrap& bootstrap,
+                                     Instance& server) {
+  ENVOY_LOG(info, "loading stats configuration");
+
+  // stats_config_ should be set before populating the sinks so that it is available
+  // from the ServerFactoryContext.
+  stats_config_ = std::make_unique<StatsConfigImpl>(bootstrap);
+  auto stats_config_impl = static_cast<StatsConfigImpl*>(stats_config_.get());
```

Comment:
@mattklein123 is it fine to down case here for the purposes of the impl specific `addSink`. Given createStatsSink takes a ServerFactoryContext as an argument, it has access to `stats_config` which is typed to `Configuration::StatsConfig`.

Reply:
Just make stats_config_ the impl type and you don't need the cast. It can be returned as the interface type.

/wait

Reply:
ah yes of course